### ISSUE TITLE
waypoint uninstall in ECS leaves tags of deleted resources

### ISSUE THREAD

**Issue by psihachina** [hashicorp/waypoint#1974]
**Describe the bug**
Repeatedly installing/uninstalling a waypoint server accumulates resources in ECS that were deleted, but are listed in the search by tag and are only visible in the [AWS CLI](https://gist.github.com/psihachina/13262db33102e29d321c979c9812822d). This results in an error on the following waypoint server uninstallation. Deleted resources are ELB listeners and EFS Access points. These resources are not found as [supported](https://docs.aws.amazon.com/ARG/latest/userguide/supported-resources.html) in the resources group in the official documentation, but these resources can tag in AWS GO SDK. Maybe is a bug in AWS.

**Steps to Reproduce**

To see the removed resources in the AWS CLI, you need to run waypoint server `install -platform=ecs` / `waypoint server uninstall -auto-approve -platform=ecs` several times

If you repeat (a lot) of these actions, there will be an [error](https://gist.github.com/psihachina/4ec4844deb230eaa681cf95e59f12122).

**Expected behavior**
Installing and uninstalling waypoint in ECS without error

**Waypoint Platform Versions**

- Waypoint CLI Version: `main`
- Waypoint Server Platform and Version:  `ecs`

**Additional context**
One solution to the problem would be to remove the tag during installation for ELB listener and EFS Access point.

Also the [difference](https://gist.github.com/psihachina/0e86c2fef48a45b9596f21b7d9714362) between serving resources by tag in the GO AWS SDK in a tag-cluttered region and almost clean

**Comment by catsby** [comment]
Hey @psihachina thank you for reporting this! 

> Deleted resources are ELB listeners and EFS Access points. 

I'm not sure what you mean by this, can you elaborate? I looked through the code and while those are tagged they, the deletion code looks for `"AWS::ElasticLoadBalancingV2::LoadBalancer"` to then query what listeners to delete. Same for EFS Access Points, we look through the tag query results and look for `"AWS::EFS::FileSystem"` to then delete it's mount targets. 

Thanks for sharing the errors in the gists as well, though they confuse me a bit. I'm not sure about the log groups because we don't allow customization there and just reference the same names at this time, I'm not sure how the group is already destroyed. Same for the cluster, but either way, I can check for their corresponding `NotFound` errors and simply continue in those cases as that's the desired result anyway. I'll open a PR for that

**Comment by psihachina** [comment]
@catsby 

> Deleted resources are ELB listeners and EFS Access points.
Here I just meant what resources are found when searching by tag in the AWS CLI.
At first, they do not interfere, and at that time the installation/removal is successful. But then, over time, the waypoint-server tag gets overwhelmed by these resources and the waypoint starts to fail when deleted, because it cannot find the resources it needs.

**Comment by catsby** [comment]
> Also, the problem is that we cannot clear the tags from these resources in any way, since they are not there.

That's interesting... I haven't yet reproduced that but I'll try using a different region. Either way we can probably just remove tags from those resources since the tag for them isn't directly used in cleanup, and they belong to a parent object that is tagged. I'll let you know what I come up with!

**Comment by catsby** [comment]
I opened https://github.com/hashicorp/waypoint/pull/1984 to better handle uninstalling the server from ECS. I'm unable to reproduce the issue described where tags show up in the CLI output but are not accessible from the AWS console. Are they visible in the AWS Resource Manager console?

**Comment by psihachina** [comment]
They ( LB listener, EFS Access point) are not visible in the AWS Resource Manager console. This is how the output in the AWS CLI looks [like for me](https://gist.github.com/psihachina/13262db33102e29d321c979c9812822d). This is after about 10 installations/uninstallations.

**Comment by psihachina** [comment]
@catsby  I looked at your PR and the use pagination function solution looks cool and should help with this problem. I was just trying to find the lost resources in this way.

**Comment by catsby** [comment]
@psihachina and those efs file systems that own the access points, do they show up in the EFS Console? What API endpoint/CLI command(s) are you using? I'm using https://docs.aws.amazon.com/ARG/latest/APIReference/API_SearchResources.html like so:

```
$ aws resource-groups search-resources --resource-query file://server-query.json
```

Where the `server-query.json` file looks like so:

```
{
  "Type": "TAG_FILTERS_1_0",
  "Query": "{\"ResourceTypeFilters\":[\"AWS::AllSupported\"],\"TagFilters\":[{\"Key\":\"waypoint-server\",\"Values\":[]}]}"
}
```

**Comment by psihachina** [comment]
@catsby 
> those efs file systems that own the access points, do they show up in the EFS Console?

Yes, show up.

>  What API endpoint/CLI command(s) are you using?

```
aws resourcegroupstaggingapi get-resources \
        --tag-filters Key=waypoint-server \
        --tags-per-page 100
```

**Comment by catsby** [comment]
I've merged #1993 which should resolve this issue. 

For the EFS resources that still show in the console, I imagine those unfortunately got orphaned when errors were encountered when destroying the ECS and CloudWatch resources (as your log showed). The server context was likely lost so `waypoint server uninstall` couldn't retry and those resources were orphaned. With #1993 the chances of that happening have been greatly reduced due to better error handling. 

Regarding the tagged items that no longer exist(?) that are showing in the CLI output, I think this is an odd thing with AWS's API here. I'm using the `resource-groups` API and I see you're using the `resourcegroupstaggingapi` API. The thing that stands out to me in the [documentation][docs] is this (emphasis mine):

> Returns all the tagged **or previously tagged** resources that are located in the specified Region for the AWS account.

I can only speculate that those resources do not exist anymore, but still qualify as "previously tagged resources". I can't say for certain though. I do see destroyed resources when I use that API call, but I don't see them in the console or in the `resource-groups` API. 

[docs]: https://docs.aws.amazon.com/cli/latest/reference/resourcegroupstaggingapi/get-resources.html
